### PR TITLE
Fix dynamic linking errors on Windows

### PR DESCRIPTION
**Pull request by yemreinci**
Dynamic linking on Windows was broken. This PR:

- Removes unused `MapAddPartitionLostListenerCodec` whose functions was declared but not defined. 
- Adds missing `HAZELCAST_API` before classes, template specializations, friend functions, std::hash/std::less specializations.
- Hides the usage of easylogger functions (enabled, info, debug etc.) in implementation files which were previously exposed through the header `ILogger.h` . This caused errors during link time because easylogger functions don't have __declspec(dllimport/dllexport) attributes specified.
- Replaces `AtomicInt` with std::atomic\<int\>. 
- Replaces `AtomicBoolean` with std::atomic\<bool\>. 
- Removes `RuntimeAvaliableProcessors` and uses `std::thread::hardware_concurrency`

### CODE REVIEW

File: hazelcast/include/hazelcast/client/serialization/pimpl/Data.h [hunk 1]
```diff
@@ -97,7 +97,7 @@ namespace std {
 
     template<>
     struct equal_to<std::shared_ptr<hazelcast::client::serialization::pimpl::Data>> {
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/serialization/pimpl/Data.h:99]: we can put export at the struct level?

File: hazelcast/include/hazelcast/util/ILogger.h [hunk 2]
```diff
@@ -102,30 +100,13 @@ namespace hazelcast {
 
             template <typename T, typename... Targs>
             void log(el::Level level, const T &value, const Targs&... fargs) {
-                if (!easyLogger->enabled(level)) {
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/util/ILogger.h:105]: why did you delete? this eliminates the composeMessage cost.

**Comment by yemreinci** [comment]
verify

**Comment by ihsandemir** [comment]
verify-windows

File: hazelcast/include/hazelcast/util/RuntimeAvailableProcessors.h [hunk 3]
```diff
@@ -63,7 +65,7 @@ namespace hazelcast {
         private:
 
             // number of available processors currently configured
-            static util::AtomicInt currentAvailableProcessors;
+            static std::atomic<int> currentAvailableProcessors;
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/util/RuntimeAvailableProcessors.h:68]: when doing this change you have to set the variable to 0 in constructor so that the behaviour is preserved.

**Review by ihsandemir — CHANGES_REQUESTED** [review]
w'll you also get rid of atomicboolean?

File: hazelcast/include/hazelcast/util/RuntimeAvailableProcessors.h (hunk 3 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/util/RuntimeAvailableProcessors.h:68]: that's a static variable, you can't initialize it here. it gets initialized in https://github.com/hazelcast/hazelcast-cpp-client/blob/a79088b7ca0cbdfed3ac1f3421c563b5f5e2c972/hazelcast/src/hazelcast/util/util.cpp#L1154.
By the way, what is the point of this variable to be an atomic? As far as I'm aware, it gets initialized only once and never changed later.
Also, to be safe, I first deleted default ctor of AtomicInt, zero initialized necessary places, compiled the code, and then finally replaced every AtomicInt with std::atomic<int>. So, there shouldn't be any atomic int variable left uninitialized.

Comment (ihsandemir) [reply hazelcast/include/hazelcast/util/RuntimeAvailableProcessors.h:68]: Ok, if it is static, and it is being initialized, it is OK. I think we can completely get rid of this class and use https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency instead.

**Comment by ihsandemir** [comment]
verify-windows

**Comment by yemreinci** [comment]
> w'll you also get rid of atomicboolean?

will do.

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

**Comment by yemreinci** [comment]
verify-linux

**Comment by yemreinci** [comment]
verify-windows

**Comment by yemreinci** [comment]
verify-windows

**Comment by yemreinci** [comment]
verify-windows

**Review by ihsandemir — APPROVED** [review]
